src/main/java/net/minestom/server/utils/collection/CollectionView.java

```diff
@@ -0,0 +1,178 @@
+package net.minestom.server.utils.collection;
+
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.function.Function;
+
+/**
+ * A CollectionView is a class which is mapped to another collection
+ * and convert every result using a given function. It is more efficient
+ * than filling a new collection every time, as long as the two types are interchangeable.
+ * <p>
+ * The view is not thread-safe.
+ *
+ * @param <E> the type that the collection should return
+ * @param <V> the type of the viewed collection
+ */
+public class CollectionView<E, V> implements Collection<E> {
+
+    private final Collection<V> collectionView;
+    private final Function<E, V> toViewFunction;
+    private final Function<V, E> toTypeFunction;
+
+    public CollectionView(@NotNull Collection<V> collectionView,
+                          @NotNull Function<E, V> toViewFunction,
+                          @NotNull Function<V, E> toTypeFunction) {
+        this.collectionView = collectionView;
+        this.toViewFunction = toViewFunction;
+        this.toTypeFunction = toTypeFunction;
+    }
+
+    @Override
+    public int size() {
+        return collectionView.size();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return collectionView.isEmpty();
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        try {
+            return collectionView.contains(toViewFunction.apply((E) o));
+        } catch (ClassCastException e) {
```

Comment:
Wouldn't this be better with an instance of check? This might throw an NPE otherwise.